lib/grpfp.gi

```diff
@@ -4042,57 +4044,69 @@ local mappow, G, max, p, gens, rels, comb, i, l, m, H, t, gen, sz,
 
   H:=[]; # indicate pseudo-size 0
   if not HasSize(G) then
-    Info(InfoFpGroup,1,"First compute size via cyclic subgroup");
-    t:=FinIndexCyclicSubgroupGenerator(G,max);
-    if t<>fail then
-      gen:=t[1];
-      Unbind(t);
-      t := NEWTC_CosetEnumerator( FreeGeneratorsOfFpGroup(G),
-            RelatorsOfFpGroup(G),[gen],true,false:
-              cyclic:=true,limit:=1+max,quiet:=true );
-    fi;
-
-    if t=fail then
-      # we cannot get the size within the permitted limits -- give up
-      return fail;
-    fi;
-    e:=NEWTC_CyclicSubgroupOrder(t);
-    if e=0 then
-      SetSize(G,infinity);
-      return fail;
-    fi;
-    sz:=e*t.index;
+    sz:=SIZE_FP_FROM_CYCLIC_INDEX(G);
     SetSize(G,sz);
-    Info(InfoFpGroup,1,"found size ",sz);
-    if sz>200*t.index then
-      # try the corresponding perm rep
-      p:=t.ct{t.offset+[1..Length(FreeGeneratorsOfFpGroup(G))]};
-      Unbind(t);
-
-      for j in [1..Length(p)] do
-        p[j]:=PermList(p[j]);
-      od;
-      H:= GroupByGenerators( p );
-      # compute stabilizer chain with size info.
-      StabChain(H,rec(limit:=sz));
-      if Size(H)<sz then
-        # don't try this again
-        comb:=Filtered(comb,i->i<>[gen]);
-      fi;
-    else
-      # for memory reasons it might be better to try other perm rep first
-      Unbind(t);
-    fi;
-
   elif Size(G)=infinity then
     return fail;
   fi;
 
   sz:=Size(G);
+
   if sz*10>max then
     max:=sz*10;
   fi;
 
+  # do we have a cyclic subgroup by which we can enumerate?
+  if HasCyclicSubgroupFpGroup(G) then
+    trial:=GeneratorsOfGroup(CyclicSubgroupFpGroup(G));
+    trial:=List(trial,UnderlyingElement);
+    Info(InfoFpGroup,1,"Try subgroup ",trial," with ",max);
+    t:=CosetTableFromGensAndRels(gens,rels,trial:silent:=true,max:=max );
+    if t<>fail and IndexCosetTab(t)>1 then
+        p:=t{[1,3..Length(t)-1]};
+        Unbind(t);
+        for j in [1..Length(p)] do
+          p[j]:=PermList(p[j]);
+        od;
+        H:= GroupByGenerators( p );
+
+        StabChain(H,rec(limit:=sz));
+
+        Info(InfoFpGroup,1,"found index ",NrMovedPoints(H)," size ",Size(H));
+
+        if Size(H)<sz then
+          # not good -- induce abelian rep
+          iso:=IsomorphismFpGroup(SubgroupNC(G,
+            List(trial,x->ElementOfFpGroup(FamilyObj(One(G)),x))
+            ):silent:=true,max:=2*max);
```

Comment:
No -- trial are elements of the free group, here we need elements of the fp group

Reply:
Sure but you then convert the elements of trial back into elements of the fp group `G` - and then form the subgroup they generate.

Reply:
OK, it would be the same, but would make the code different from the similar bit below. (There is a lot of further opportunity for clean-up, but since this is code I rarely use, but a lot of users do, I need to be careful about introducing errors).